cpu/stm32/periph/eth.c

```diff
@@ -254,16 +256,102 @@ static int stm32_eth_get(netdev_t *dev, netopt_t opt,
 static void _timer_cb(void *arg)
 {
     netdev_t *dev = (netdev_t *)arg;
+    uint8_t state = LINK_STATE_DOWN;
     if (_get_link_status()) {
-        _link_state = LINK_STATE_UP;
+        state = LINK_STATE_UP;
+    }
+
+    if ((_link_state & LINK_STATE) != state) {
+        /* link state changed, notify upper layer */
+        _link_state = state;
         dev->event_callback(dev, NETDEV_EVENT_ISR);
     }
-    else {
-        _link_state = LINK_STATE_DOWN;
-        xtimer_set(&_link_status_timer, STM32_ETH_LINK_UP_TIMEOUT_US);
+
+    xtimer_set(&_link_status_timer, STM32_ETH_LINK_UP_TIMEOUT_US);
+}
+#endif /* IS_USED(MODULE_STM32_ETH_LINK_UP) */
+
+static bool _phy_can_negotiate(void)
+{
+    return (_mii_reg_read(MII_BMSR) & MII_BMSR_HAS_AN);
+}
+
+#if IS_USED(MODULE_STM32_ETH_AUTO)
+static void _complete_auto_negotiation(void)
+{
+    /* first, wait until auto-negotiation really has completed */
+    uint16_t bmsr;
+
+    do {
+        bmsr = _mii_reg_read(MII_BMSR);
+        if (!(bmsr & MII_BMSR_LINK)) {
+            /* disconnected during auto-negotiation */
+            return;
+        }
+    } while (!(bmsr & MII_BMSR_AN_DONE));
+    DEBUG("[stm32_eth] PHY auto-negotiation completed, PHY link up\n");
+    /* Get current MACCR state without speed config */
+    uint32_t maccr = ETH->MACCR & ~(ETH_MACCR_FES | ETH_MACCR_DM);
+    /* stupidly, there is seemingly no way to get current connection speed
+     * and duplex mode. But we can deduce it from our advertised capabilities
+     * and the link partner abilities */
+    uint16_t adv = _mii_reg_read(MII_ADVERTISE);
+    uint16_t lpa = _mii_reg_read(MII_LPA);
+
+    if ((adv & MII_ADVERTISE_100) && (lpa & MII_LPA_100)) {
+        /* 100 Mbps */
+        maccr |= ETH_MACCR_FES;
+        if ((adv & MII_ADVERTISE_100_F) && (lpa & MII_LPA_100_F)) {
+            /* full duplex */
+            maccr |= ETH_MACCR_DM;
+        }
     }
+    else if ((adv & MII_ADVERTISE_10_F) && (lpa & MII_LPA_10_F)) {
+        /* full duplex */
+        maccr |= ETH_MACCR_DM;
+    }
+    DEBUG("[stm32_eth] %s Mbps %s duplex \n",
+          (maccr & ETH_MACCR_FES) ? "100" : "10",
+          (maccr & ETH_MACCR_DM) ? "full" : "half");
+    ETH->MACCR = maccr;
+}
+#endif /* IS_USED(MODULE_STM32_ETH_AUTO) */
+
+static void _setup_phy(void)
+{
+    DEBUG("[stm32_eth] Reset PHY\n");
+    /* reset PHY */
+    _mii_reg_write(MII_BMCR, MII_BMCR_RESET);
+    /* wait till PHY reset is completed */
+    while (MII_BMCR_RESET & _mii_reg_read(MII_BMCR)) { }
```

Comment:
```suggestion
    while (MII_BMCR_RESET & _mii_reg_read(MII_BMCR)) {}
```

Reply:
I think this is consistent with the other code in RIOT. I'm personally fine with either.

Reply:
I just ran a `grep` and can confirm that `{}` is indeed more widely used than `{ }`. So let me fix this.